### ISSUE TITLE
bug(builders): semantic release fails when releasing multiple projects

### ISSUE THREAD

**Issue by javisank** [ng-easy/platform#631]
Hi! I'm building a repository with Nx for library development, in which many libraries depend on each other, or in parent libraries (all inside said repository).
When making changes on a single library, everything works just fine. But if that library has dependencies, or if I made a change spanning more than one library, then it only releases some (usually the ones that don't have dependencies, or the ones that their dependencies did not change). I'll explain further:

I use the project.json file to configure every project' targets, and this is my release config:
```
"release": {
      "executor": "@ng-easy/builders:semantic-release",
      "options": {
        "force": false,
        "dryRun": false,
        "npm": true,
        "github": true,
        "githubSuccessComment": false,
        "releaseCommitMessage": "chore(release): :package: ${project}@${nextRelease.version} [skip ci]",
        "branches": [
          "main",
          {
            "name": "beta",
            "prerelease": true
          }
        ],
        "changelog": true
      }
    }

```

Github Actions CI (for beta/pre-release) release config:
```
   - name: Release
     uses: mansagroup/nrwl-nx-action@v2
     with:
        targets: release
        affected: true
     env:
        CI: true
```

An example of this:
- errors (no dependencies)
- auth (depends on errors)
- storage (depends on errors)

I make changes on auth and storage, and it runs this:
nx -- affected --target=release --base=55a4cfcb2602e870f8c81f7ebb218dfd901abc42 --head=0bceb3d4ac88f2842be6d247e172ffc6d9879989

1. No new release for the "errors" project --> **This is OK**
2. [6:54:13 PM] [semantic-release] › :information_source:  Start step "prepare" of plugin "Inline plugin"
[6:54:13 PM] [semantic-release] [Inline plugin] › :information_source:  Nx Update Package Version: Prepare
[6:54:13 PM] [semantic-release] [Inline plugin] › :information_source:  Updating package version on files:
[6:54:13 PM] [semantic-release] [Inline plugin] › :information_source:  - packages/libs/module/storage/CHANGELOG.md
[6:54:13 PM] [semantic-release] [Inline plugin] › :information_source:  - packages/libs/module/storage/package.json
[6:54:13 PM] [semantic-release] [Inline plugin] › :information_source:  Found 2 file(s) to commit
[6:54:15 PM] [semantic-release] › :heavy_multiplication_x:  Failed step "prepare" of plugin "Inline plugin"
[6:54:15 PM] [semantic-release] › :heavy_multiplication_x:  An error occurred while running semantic-release: Error: Command failed with exit code 1: git push --tags https://x-access-token/:[secure]@github.com/my-organization/npm-packages HEAD:beta
To https://github.com/my-organization/npm-packages
 ! [remote rejected] HEAD -> beta (cannot lock ref 'refs/heads/beta': is at 9d492ec1781d1921a27974126006a62a1a429ec0 but expected 0bceb3d4ac88f2842be6d247e172ffc6d9879989)
**This is not OK**, and I do not understand the difference in the refs (I think maybe it is because of commits made by semantic release for bump made in another project.
3. The "auth" project was released with version 1.0.0-beta.2 --> **This is OK**

What I want to achieve is having the possibility to merge a PR into a branch, and having every changed library (and the ones that depend on it) released and published. I hope I made myself clear and that somebody could help me.

**Comment by samuelfernandez** [comment]
@javisank thanks for reporting this issue! It looks very much similar to #599. Let's see if we can troubleshoot here a bit more here.

First of all, the release builder should run for all projects, since otherwise there is no way to cascade dependent projects and release them. You can read the full explanation [here](https://github.com/ng-easy/platform/issues/599#issuecomment-1174961256) and also in the [README](https://github.com/ng-easy/platform/tree/main/libs/builders/src/semantic-release#faq) under question `How can I configure GitHub CI to run automatically the release?`.

So basically, you should use remove `affected: true` from your release action.

Having that said, the error you describe has something else going on, when it fails for not being in sync with the head branch in the remote repo. Current theory is that while the semantic release process was running there was another commit into `beta` branch, namely, `9d492ec1781d1921a27974126006a62a1a429ec0`. Would you be able to check the timing of when the release process executed, and the time when that commit was applied to `beta` branch? If that is the case, that error is expected, see https://github.com/semantic-release/semantic-release/issues/993 and https://github.com/semantic-release/semantic-release/issues/1208 (I'm assuming that you are using this in a private repo so can't share the full logs, however that would be useful).

In addition to that, I'd like to encourage you to try two [hidden flags](https://github.com/ng-easy/platform/commit/418abc8cff45cb68b6a1d74b130e324f3ee4ab65#diff-5548481a65b93025b3019ff31ff108dba3052bd80970e02073e7251f0928b563R56) to enhance troubleshooting. Could you please add to the options of your release executor this?

```json
{
  "options": {
    "verbose": true,
    "forceGitPullRebase": true
  }
}
```

Let's see if this helps to either solve the problem or shed light on the underlying issue 😄 

cc @johannesschobel

**Comment by johannesschobel** [comment]
Dear @samuelfernandez , @javisank ,
i guess, this is indeed the same issue as i have encountered.

i think, the issue arises, because the first package that is released creates a commit that is immediately pushed to the `main` branch. The next release steps internally checks, if the latest fetched commit is the same as the current head of the branch to be released from. And this check fails. And that is the message you encounter.

@samuelfernandez already provided a fix, however, i was not able to try it because i am quite busy at the moment. maybe you can try it?

All the best

**Comment by samuelfernandez** [comment]
> i think, the issue arises, because the first package that is released creates a commit that is immediately pushed to the `main` branch. The next release steps internally checks, if the latest fetched commit is the same as the current head of the branch to be released from. And this check fails. And that is the message you encounter.

This is the normal workflow. You can see in this [run](https://github.com/ng-easy/platform/runs/7375431108?check_suite_focus=true) how "base" libraries are built, then the release creates a commit to update dependent libraries, and it all succeeds. As new libraries get released, they should compare the new commit with the remote, and both should be in sync.

**Comment by mi-kei-la** [comment]
Hi! First of all thank you both for answering so quickly! I'm working with @javisank on this, and today we spent the day making a series of tests (I will attach all the logs for each run and explain the variables we tweaked on each). Hopefully that gives you a better insight, since we haven't been able to make it work as expected yet.

First of all, we removed the flag **affected** from all the following runs, and installed @ng-easy/builders@9.0.8. We tested the options you first mentioned (verbose + forceGitPullRebase), and it failed unexpectedly with a message that GitHub could not authenticate.

After that, we started to make a series of controlled changes to see what happened.

What we found out is that setting `"forceGitPullRebase": true` in the release options leads us to the following error:
`Command failed with exit code 128: git ls-remote --heads https://github.com/umahealth/BE-npm-packages beta
fatal: could not read Username for 'https://github.com/': No such device or address` which was solved by setting the checkout step like this
```
      - uses: actions/checkout@v3
        with:
          fetch-depth: 0
          token: ${{ secrets.UMA_GITHUB_ADMIN_TOKEN }}
          persist-credentials: true # Needed so that semantic release can use the admin token
```
[forceGitPullRebase: true, persist-credentials: true](https://github.com/ng-easy/platform/files/9162545/logs_80.zip): We got this error:
```
Git rebase:
Command failed with exit code 128: git pull --rebase
fatal: Cannot rebase onto multiple branches.
```
------------

[Another test: persist-credentials: false, forceGitPullRebase: false](https://github.com/ng-easy/platform/files/9162616/logs_82.zip)
```
options: {
        "verbose": true,
        "forceGitPullRebase": false
}
```
This is the publish.yaml config for that run
```
      - uses: actions/checkout@v3
        with:
          fetch-depth: 0
          persist-credentials: false # Needed so that semantic release can use the admin token
```
Error:
```
Command failed with exit code 128: git ls-remote --heads https://github.com/umahealth/BE-npm-packages beta
fatal: could not read Username for 'https://github.com/': No such device or address
```
--------------
[And finally, we went full circle: persist-credentials: true, forceGitPullRebase: false](https://github.com/ng-easy/platform/files/9162658/logs_84.zip)
And it failed with the original error message:
```
[8:35:11 PM] [semantic-release] › ✖  An error occurred while running semantic-release: Error: Command failed with exit code 128: git commit -m chore(release): :package: errors@1.0.0-beta.1 [skip ci]
→ No staged files match any configured task.
fatal: cannot lock ref 'HEAD': is at 33133f9e644409e32437cd13dfd6c096d5647365 but expected bc20a2c52125e6b0c5de8589e5ec85eac50f8aab
```

These are the commits that are being compared:
![my commit, and the semantic-release commit](https://user-images.githubusercontent.com/66849994/180311689-7d0b23a5-f558-43d1-8192-fa49b53e0b8f.png)
As you can see, HEAD is at the latest semantic-release commit, but it expected the commit I made.

Again, thank you for your time and collaboration, we don't have any new ideas of what to test to make this work so we are truly grateful. We hope you have a great week c:

**Comment by samuelfernandez** [comment]
@mi-kei-la Thanks for this detailed explanation, it is really helpful.

Side note: from your logs I've noticed that the `nx release` command is run in parallel:

```
2022-07-21T20:35:02.2358193Z ##[group]Run mansagroup/nrwl-nx-action@v2
2022-07-21T20:35:02.2358552Z with:
2022-07-21T20:35:02.2358830Z   targets: release
2022-07-21T20:35:02.2359175Z   affected: false
2022-07-21T20:35:02.2359465Z   parallel: true
2022-07-21T20:35:02.2359746Z   all: false
2022-07-21T20:35:02.2360035Z   maxParallel: 3
2022-07-21T20:35:02.2360272Z   nxCloud: false
```

That is not expected, maybe you need to enforce disabling the parallel run.

Let's go back to the main issue. The release commit updates the changelog and dependent packages. It seems that the commit is created, but fails when setting that to HEAD. So it points towards a permissions issue. 

Official documentation for [`@semantic-release/git`](https://github.com/semantic-release/git#git-authentication) docs explains that:

> The Git user associated with the [Git credentials](https://github.com/semantic-release/semantic-release/blob/master/docs/usage/ci-configuration.md#authentication) has to be able to push commit to the [release branch](https://github.com/semantic-release/semantic-release/blob/master/docs/usage/configuration.md#branch).
> When configuring branches permission on a Git hosting service (e.g. [GitHub protected branches](https://help.github.com/articles/about-protected-branches), [GitLab protected branches](https://docs.gitlab.com/ee/user/project/protected_branches.html) or [Bitbucket branch permissions](https://confluence.atlassian.com/bitbucket/branch-permissions-385912271.html)) it might be necessary to create a specific configuration in order to allow the semantic-release user to bypass global restrictions. For example on GitHub you can uncheck "Include administrators" and configure semantic-release to use an administrator user, so the plugin can push the release commit without requiring [status checks](https://help.github.com/articles/about-required-status-checks) and [pull request reviews](https://help.github.com/articles/about-required-reviews-for-pull-requests).

So, in order to make it work you should take this into consideration:

* Use `persistCredentials: false` in the checkout repo step:

```yml
      - name: Checkout repo
        uses: actions/checkout@v2.3.4
        with:
          fetch-depth: 0
          persist-credentials: false # Needed so that semantic release can use the admin token
```

* Use a release token with repo permissions in the release step:

```yml
      - name: Release
        uses: mansagroup/nrwl-nx-action@v2
        with:
          targets: release
          nxCloud: 'true'
        env:
          CI: true
          NPM_TOKEN: ${{ secrets.NPM_TOKEN }}
          GITHUB_TOKEN: ${{ secrets.RELEASE_TOKEN }} # Personal access token with repo permissions
```

![image](https://user-images.githubusercontent.com/19343955/180603330-e692fd16-9226-46ed-b928-020df6baaa8b.png)

* In your branch protection rules, enable forced push and not enforce administrators (as explained in the documentation)

![image](https://user-images.githubusercontent.com/19343955/180603401-a2960fef-62fa-47c5-880c-053dca5f1757.png)

I hope this works! 😄

**Comment by javisank** [comment]
Hello, I think I'll give it a shot on monday. One question tough: what's the reason behind asking for a linear history?
Thank you and have a nice weekend.

**Comment by samuelfernandez** [comment]
> Hello, I think I'll give it a shot on monday. One question tough: what's the reason behind asking for a linear history?
> 
> Thank you and have a nice weekend.

I think it's not needed, just push rights to HEAD

**Comment by johannesschobel** [comment]
a linear history is more readable than having multiple branches running all towards `main` and merging eventually. It makes the commit history more consistent, i think.

however, whether `merging` or `rebasing` is better is some kind of a religious debate, i guess, haha

**Comment by mi-kei-la** [comment]
@samuelfernandez thank you for the detailed answer!! Sadly, it's still not working. We configured the repo as you say, set parallel in false, double-checked that the token had permisions, aaaand encountered this error:
```
>  NX   Command failed with exit code 128: git ls-remote --heads https://github.com/umahealth/BE-npm-packages beta

   fatal: could not read Username for 'https://github.com/': No such device or address
   Pass --verbose to see the stacktrace.
```
This is not the first time that happens, and it's always when we set `persist-credentials: false`. When we set it to `true` though, we encounter again the different heads error. [This is the full log, in case it gives you any sort of answer](https://github.com/ng-easy/platform/files/9184063/logs_92.1.zip)

------

Second run. We enabled `forceGitPullRebase: true` and set `persist-credentials: true` to avoid the latest error. And this is the new error:
```
>  NX   Command failed with exit code 128: git pull --rebase

   warning: fetch updated the current branch head.
   fast-forwarding your working tree from
   commit 8cf[70](https://github.com/umahealth/BE-npm-packages/runs/7506762987?check_suite_focus=true#step:6:72)a1d586fbdd44b77654201c880699c983220.
   fatal: Cannot rebase onto multiple branches.
   Pass --verbose to see the stacktrace.
```
[Full log c:](https://github.com/ng-easy/platform/files/9184130/logs_96.zip)

-----

Third run: `githubSuccessComment: false` (to avoid noisy emails) and `persist-credentials: false`. Again the no username error:
```
>  NX   Command failed with exit code 128: git ls-remote --heads https://github.com/umahealth/BE-npm-packages beta

   fatal: could not read Username for 'https://github.com/': No such device or address
   Pass --verbose to see the stacktrace.
```
[Third run log](https://github.com/ng-easy/platform/files/9184154/logs_97.zip)

No more runs for the day. Thank you

**Comment by samuelfernandez** [comment]
Thanks for the detailed explanation, I'll try to take a look to it tomorrow. This is the documentation that should be followed https://github.com/semantic-release/semantic-release/blob/master/docs/recipes/ci-configurations/github-actions.md

Leaving here for reference, in case you spot anything that is missing.

**Comment by samuelfernandez** [comment]
Related to https://github.com/semantic-release/git/issues/196

**Comment by samuelfernandez** [comment]
@mi-kei-la after checking your logs, I have a feeling that the two additional flags I added are causing unexpected side effects and errors. Let's try to remove them again from your configuration ( `"verbose": false, "forceGitPullRebase": false`)

After re-reading all documentation from `semantic-release/git`, these are the key configuration points that should make it work:

https://github.com/ng-easy/platform/issues/631#issuecomment-1193111465

* Use `persistCredentials: false`
* Use personal token with full permissions. Create a secret for it different than `GITHUB_TOKEN` and pass it to the input. Could you share your full configuration in the `yml` workflow for the release job? In the example I put above I used `RELEASE_TOKEN`.
* It should have `repo` permissions, and be owner of the repo

If you google a bit on "how to push to protected branch github action" you can find these explanations:

> PAT user permissions
> 
> The user that the PAT represents MUST have "admin" permission to the repository in order to handle protected branches: determine which checks are running/finished and to toggle the "require review"-protection.
> 
> If the PAT represents the repository owner, there are no issues, however, if the PAT represents a collaborator, the collaborator MUST be given the "Admin" role. This can be done under the "Settings" tab in the repository and then going to "Collaborators & teams". To understand what the "Admin" role allows the user to do, you can see the "Repository roles" page, which is also found under the "Settings" tab in the repository.

Some additional explanations than could help:

* https://gullerya.medium.com/github-actions-automated-commit-with-yet-required-checks-13eebf0d3c70
* https://solveforum.com/forums/threads/solved-how-to-push-to-protected-branches-in-a-github-action.882184/#post-882574

Also, it seems it is suggested to check the `repository.url` in `package.json` as described [here](https://github.com/semantic-release/git/issues/196#issuecomment-590721885)

You can also check [this thread](https://github.com/semantic-release/git/issues/196#issuecomment-702901192), which gives a more detailed explanation.

Let's see if after double checking this the errors are gone, or at least it would help to identify the underlying problem. I feel we are getting closer! 😄 Again, thanks for the patience

**Comment by tomavic** [comment]
Hi folks,

I had the same issue, I had simple project here 

https://github.com/tomavic/enigma-nx-angular

with minimal setup for two libs and one app

I don't know if the problem is with releasing two libs at the same time or is it just normal to do this.

Should I separate GH action into multiple parallel jobs :/ 

I've checked your repo itself but I can see there is a bit complicated setup for semantic release 😸

**Comment by samuelfernandez** [comment]
> Hi folks,
> 
> I had the same issue, I had simple project here
> 
> https://github.com/tomavic/enigma-nx-angular
> 
> with minimal setup for two libs and one app
> 
> I don't know if the problem is with releasing two libs at the same time or is it just normal to do this.
> 
> Should I separate GH action into multiple parallel jobs :/
> 
> I've checked your repo itself but I can see there is a bit complicated setup for semantic release 😸

Hey @tomavic thanks for reaching out! In theory it should be supported. However, this solution has some scalability problems. I'd suggest trying out this plugin https://github.com/jscutlery/semver which might offer a better solution. Sorry for the inconvenience.

**Comment by tomavic** [comment]
Thanks @samuelfernandez for replay. I really want to discuss this with someone who faces this kind of issues as this does not seem to be popular topic a among the public :D 

I believe that the main problem is with the @semantic-release/git plugin itself not the way semantic-release works and this is even worse 😐 

I've searched for many solutions for versioning packages and apps and I know Semver but didn't try it before. (Maybe I should give it a try)

The reason why I did choose semantic-release for single repo. is the variety of plugins and ability to customise, however I did spend a lot of time  to configure it 😢 but at the end I collected my favourite configs and used them across many projects.

When I started to migrate to monorepo, I wanted to use my semantic-release lovely configs again and I shocked when I found that there is no intent to support monorepo and you gotta find help from community.

I've tried https://github.com/TheUnderScorer/nx-semantic-release in my repo here https://github.com/tomavic/enigma-nx-angular/tree/feat/nx-semantic-release and it seems to be a good solution though it is limited for modification and If someone wanted to add custom configs he has to clone it or ask the maintainer and wait for him to replay 😒
